### DISCUSSION TITLE
Adding leading zeros to string don't get saved to database

### DISCUSSION THREAD

**Discussion by vixducis** [laravel/framework#49687] [Q&A]
### Laravel Version

10.40.0

### PHP Version

8.2

### Database Driver & Version

MySQL 8.0.30 on CentOS 7 (x86)

### Description

Let's say we have an eloquent model `Test`, that has a string property `prop`. A database record exists where `prop` contains the string value `"1234"`. We load this record using eloquent models, and update the value to contain leading zeros like this:
```php
$model->prop = "01234";
$model->save();
```
In this case, the database value will not be updated. This is caused by the following code in `src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php` on line 2103:
```php
return is_numeric($attribute) && is_numeric($original)
            && BigDecimal::of($attribute)->isEqualTo($original);
```
Since the string values do pass `is_numeric` and their number counterparts are equal, the property isn't considered dirty and the field isn't updated. These string values might represent totally different things in the database and shouldn't be considered equal in all cases.

I understand changing this might impact existing codebases, but is there any way we can provide an easier escape hatch than overwriting `HasAttributes::originalIsEquivalent` (as overwriting core functionality tends to break in the future)?

### Steps To Reproduce

```php

use Illuminate\Database\Eloquent\Model;

/**
 * @property string $prop
 */
class Test extends Model {
  protected $fillable = ['prop'];
}

$test = new Test(['prop' => '1234']);
$test->save();

$test->prop = '01234';
$test->save(); // saving doesn't happen because prop isn't considered dirty
```

**Comment by rojtjo** [comment]
This seems to be happening because no explicit cast has been defined and it has to guess the data type. I believe you can workaround this by explicitly casting to `string`:

```php
class Test extends Model {
    protected $fillable = ['prop'];
    protected $casts = ['prop' => 'string'];
}
```

See: https://github.com/laravel/framework/blob/10.x/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L2092-L2094

**Reply by rojtjo** [reply]
This might actually be related to #49504 which is being reverted in #49702

**Comment by vixducis** [comment]
I hadn't considered that, this solves my issue indeed. Thanks!
